### PR TITLE
Custom convictor for stale stats_teads_long data

### PR DESCRIPTION
**Pull request by arvy**
how to run tests: https://github.com/ebuzzing/cassandra-util/pull/2

If you don't have your Datastax credentials on hand (which are needed to download the DSE packages), I can give you mine

### CODE REVIEW

File: deleting-compaction-strategy/src/test/scala/tv/teads/cassandra/compaction/LongStatsHourlyConvictorSpec.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2016 ProtectWise, Inc.
```
Comment (arvy) [inline deleting-compaction-strategy/src/test/scala/tv/teads/cassandra/compaction/LongStatsHourlyConvictorSpec.scala]: remove protectwise legal

File: deleting-compaction-strategy/src/test/scala/tv/teads/cassandra/compaction/LongStatsHourlyConvictorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      truncate(s"$ks.$table")
+      setupData(s"$ks.$table",
+        "dim" || "dimid" || "level" || "date" || "event"  || "value"
+          | "some_dim"    !! "999" !!  "day"  !! "20160701"   !! "fill" !! 1l // kept b/c not hour
+          | "some_dim"    !! "999" !! "hour"  !! "2016071500" !! "fill" !! 1l // X
+          | "some_dim"    !! "999" !! "hour"  !! "2016073012" !! "fill" !! 1l // X
+          | "some_dim"    !! "999" !! "hour"  !! "2016080123" !! "fill" !! 1l // X
+          | "some_dim"    !! "999" !! "hour"  !! "2016081006" !! "fill" !! 1l // X
+          | "some_dim"    !! "999" !!  "day"  !! "20170130"   !! "fill" !! 1l // kept b/c not hour
+          | "some_dim"    !! "999" !! "hour"  !! "2017090112" !! "fill" !! 1l // kept b/c post-2017
```
Comment (arvy) [inline deleting-compaction-strategy/src/test/scala/tv/teads/cassandra/compaction/LongStatsHourlyConvictorSpec.scala]: also test boundaries

File: deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+     * @return
+     */
+    protected boolean isStatStale(ByteBuffer value)
+    {
+
+        try {
+            String period = ByteBufferUtil.string(value);
+            boolean evict = period.length() == 10 // hour
+                    && period.compareTo("20170101") < 0; //pre-2017
+            if( evict ) logger.debug("Evicting date={}", period);
```
Comment (arvy) [inline deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: maybe this should be TRACE?

Comment (arvy) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: in any case, i should check if the desired level is even enabled, before logging

File: deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * Returns true if value is a pre-2017 hourly stat
+     *
+     * @return
+     */
+    protected boolean isStatStale(ByteBuffer value, AbstractType<?> comparator)
+    {
+        String period = comparator.getString(value);
+        boolean evict = period.length() == 10 // hour
+                && period.compareTo("20170101") < 0; //pre-2017
```
Comment (arvy) [inline deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: it would be more proper to use the C* comparator

File: deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
+package tv.teads.cassandra.compaction;
+
+import org.apache.cassandra.config.ColumnDefinition;
+import org.apache.cassandra.cql3.ColumnIdentifier;
+import org.apache.cassandra.db.*;
+import org.apache.cassandra.db.columniterator.OnDiskAtomIterator;
```
Comment (arvy) [inline deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: redundant imports

File: deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Returns true if value is a pre-2017 hourly stat
+     *
+     * @return
+     */
+    protected boolean isStatStale(ByteBuffer value, AbstractType<?> comparator)
+    {
+        String period = comparator.getString(value);
+        boolean evict = period.length() == 10 // hour
+                && comparator.compare(value, ByteBufferUtil.bytes("20170101")) < 0; //pre-2017
```
Comment (bjaglin) [inline deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: extract constant to avoid re-computation

File: deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    public boolean shouldKeepPartition(OnDiskAtomIterator key) {
+        return true;
+    }
+
+    @Override
+    public boolean shouldKeepCluster(OnDiskAtomIterator partition, Composite name) {
+
+        for (ColumnDefinition def : cfs.metadata.clusteringColumns())
+        {
+            if ( def.name.bytes.equals(ByteBufferUtil.bytes("date"))  )
```
Comment (bjaglin) [inline deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: extract constant to avoid re-computation

File: deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        }
+        return true;
+    }
+
+    /**
+     * Returns true if value is a pre-2017 hourly stat
+     *
+     * @return
+     */
+    protected boolean isStatStale(ByteBuffer value, AbstractType<?> comparator)
```
Comment (bjaglin) [inline deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: if you only need a comparator, reflect that on the type signature

File: deleting-compaction-strategy/src/test/scala/tv/teads/cassandra/compaction/LongStatsHourlyConvictorSpec.scala [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+          | "some_dim"    !! "999" !! "hour"  !! "2016081006" !! "fill" !! 1l // X
+          | "some_dim"    !! "999" !! "hour"  !! "2016123123" !! "fill" !! 1l // X
+          | "some_dim"    !! "999" !! "hour"  !! "2017010100" !! "fill" !! 1l // kept b/c post-2017
+          | "some_dim"    !! "999" !!  "day"  !! "20170130"   !! "fill" !! 1l // kept b/c not hour
+      )
+
+      compactAllFiles(ks, table)
+
+      val result = Await.result(cql"SELECT count(*) AS c FROM ${Inline(ks)}.${Inline(table)}".execute(), Duration.Inf)
+      "right number of records after" ==> (result.one().getLong("c") mustEqual 3)
```
Comment (bjaglin) [inline deleting-compaction-strategy/src/test/scala/tv/teads/cassandra/compaction/LongStatsHourlyConvictorSpec.scala]: assert on the actual lines left rather than the number (giving them unique names)

File: deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java (hunk 8 above)
Comment (arvy) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: that's true - I copied this from the protectwise code - but looking at the hierarchy it seems the Comparator interface is sufficient

Comment (arvy) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: Actually, I need AbstractType for the `.getString(..)` method

Comment (bjaglin) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: I see.. wouldn't it be safer to pass the `level` value rather than relying on the length of something which is meant to be used for logging https://github.com/odnoklassniki/apache-cassandra/blob/master/src/java/org/apache/cassandra/db/marshal/AbstractType.java#L42-L43 ?

Comment (arvy) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: Great question - the answer is we can't, because we don't the values of the partition key. And because the partition key is composite (made up of 3 parts: dim, dimid, level) we can't do this: `hash("hour")==hashedLevelValueFromSstable`

Comment (arvy) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: I did see that comment - not sure why it's "only suitable for logging". Alternatively I could translate the value to string with ByteBufferUtils.string(byteValue)

Comment (bjaglin) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: Oh, I see, so the `Composite` doesn't include anything from the partition key?

Comment (arvy) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: No, `Composite` is actually an abstraction for a "composite" column (a column that could potentially include other columns, as in the deprecated SuperColumn feature).
In the pre-3.x sstable format we only get the hashed partition key. If `stats_teads_long` had a simple (single) partition key, we could check it for equality by simply hashing the key we're comparing.

Comment (bjaglin) [reply deleting-compaction-strategy/src/main/java/tv/teads/cassandra/compaction/LongStatsHourlyConvictor.java]: I understand, thanks for the explanation

**Comment by arvy** [comment]
retest

**Comment by arvy** [comment]
tests cannot run in jenkins, because we need a new docker container that contains both SBT and python. They run fine locally
